pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/NoDictionarySingleColumnGroupKeyGenerator.java

```diff
@@ -163,84 +164,201 @@ public int getCurrentGroupKeyUpperBound() {
 
   @Override
   public Iterator<GroupKey> getUniqueGroupKeys() {
-    return new GroupKeyIterator(_groupKeyMap);
+    switch (_dataType) {
```

Comment:
This may sound like an over-optimization but adding type specific GroupKeyGenerator has the potential to add overhead for runtime dispatch which Java doesn't handle well.

What is the advantage of having a type specific next method like the following for Double

```
@Override
    public GroupKey next() {
      Double2IntMap.Entry entry = _iterator.next();
      _groupKey._groupId = entry.getIntValue();
      _groupKey._stringKey = Double.toString(entry.getDoubleKey());
      return _groupKey;
    }
```
v/s the existing generic method

```
@Override
    public GroupKey next() {
      Map.Entry<Object, Integer> entry = _iterator.next();
      _groupKey._groupId = entry.getValue();
      _groupKey._stringKey = entry.getKey().toString();
      return _groupKey;
    }
```

Both are doing toString()

Reply:
The main difference is from the entry-set and iterator implementation of the fastutil maps. For an example, `Int2IntMap.entrySet()` is marked deprecated and the class suggest using the type-specific method instead. Also, we are able to use the `fastIterator()` provided by the `FastEntrySet`, and also the unboxed values to reduce the garbage, and the unnecessary boxing/unboxing.